helix-core/src/main/java/org/apache/helix/task/JobDispatcher.java

```diff
@@ -317,6 +316,9 @@ private ResourceAssignment computeResourceMapping(String jobResource,
     // For delayed tasks, trigger a rebalance event for the closest upcoming ready time
     scheduleForNextTask(jobResource, jobCtx, currentTime);
 
+    // If the assigned instance is no longer live, so mark it as DROPPED in the context
+    markPartitionsWithoutLiveInstance(jobCtx, allPartitions, liveInstances);
```

Comment:
This looks like an override of processing result... Can we have the logic in somewhere in scheduling task? This is more like a patching after scheduling computation. It waste the scheduling computation if instance is not live.

Reply:
We have this logic in the handleAdditionalTasks in filterTask method. Since that part of the code only hits when TargetState is equal to start, I moved the logic here so in all conditions we evaluate it. If the instance is not live, we still need to schedule the task if the task needs to be scheduled on other instances. If the instance is not live, then we do not see the task in current states which means we would no process the task in this case.

Reply:
I understand the complexity of doing the check. So my question is when could this happening? We need to think about to put to the right place. Add here works, but it looks like a "saving action" without understand any assignment logic. I would prefer to understand the scenario and move the check logic for assignment there.

Reply:
It happens when workflow/job is being stopped and while the message is being sent and not processed yet (or the controller has not seen the effect yet) then the live instance is gone. In this case, the job will be stuck in the STOPPING state and the controller does not have a way to mark the task and job to be STOPPED. Note that all new assignment is happening when TargetState is START in handleAdditionTaskAssigment. If not, then assignment logic does not exist and will not be considered. We need to somehow manage this case before that. Note that when the live instance is gone, the controller does not see anything on the instance. I cannot find a better way to achieve this. If you know how it can be resolved, please let me know and I can change it accordingly. Also if you look at handleAdditionalTaskAssignment.filterTask method, it is the same code. I basically changed the place of this code and move it earlier in the pipeline.

Reply:
If this is the only scenario, my suggestion is moving the logic to following check in JobDispatcher. It checks whether we can determine a job is already stopped or not. It is also the right time that we check whether the instance assigned with tasks are gone or not. If it is gone, we can mark it as DROPPED.

  public static boolean checkJobStopped(JobContext jobContext) {
    for (int partition : jobContext.getPartitionSet()) {
      TaskPartitionState taskState = jobContext.getPartitionState(partition);
      if (taskState == TaskPartitionState.RUNNING) {
        return false;
      }
    }
    return true;
  }

Reply:
Moved to new place as suggessted.